core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2098,11 +2098,55 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    val n         = as.size
+    val resultArr = new Array[Any](n)
+    val resultList: ZIO[R, E, List[B]] = for {
+      parentId <- ZIO.fiberId
+      latch    <- CountdownLatch.make(n)
+      failed   <- Promise.make[Nothing, Int]
+      cause    <- Ref.make[Cause[E]](Cause.empty)
+      fibers <- as
+                 .foldLeft[(Int, URIO[R, List[(Int, Fiber[E, B])]])]((0, URIO.succeed(Nil))) { (r, a) =>
+                   r match {
+                     case (i, acc) =>
+                       def handleExit(exit: Exit[E, B]): URIO[R, Any] = exit.foldM(
+                         c => cause.update(cs => cs && c) *> failed.succeed(i),
+                         b => ZIO.effectTotal(resultArr(i) = b)
+                       )
+
+                       val appended = for {
+                         fs <- acc
+                         f <- ZIO
+                               .bracketExit(
+                                 ZIO.succeed(a),
+                                 (_: A, exit: Exit[E, B]) => handleExit(exit).ensuring(latch.countDown),
+                                 fn
+                               )
+                               .fork
+                       } yield (i, f) :: fs
+
+                       (i + 1, appended)
+                   }
+                 }
+                 ._2
+      interrupter = failed.await
+        .flatMap(
+          i =>
+            ZIO.foreach(fibers) {
+              case (j, f) if i != j => f.interruptAs(parentId).unit
```

Comment:
Given that interrupting a fiber that has already terminated is a noop, you could just write
```scala
ZIO.foreach_(fibers)(_.interruptAs(parentId))
```
and only keep track of the fibers without their index.

Reply:
It's not guaranteed that it will have terminated when you get to that point (but it will indeed have finished anything that is interruptible anyway). I wrote it like that for clarity of intention. I guess a comment would do just as well for that though.

Reply:
What I meant was that `i == j => f.interruptAs(parentId).unit == ZIO.unit`, as `failed` is set to the index of the first failing fiber, so I think this one is guaranteed.